File: frontend/src/api/geneCatalogue.js
/**
 * Gene Catalogue API client.
 *
 * The Gene Catalogue is a read-only aggregation layer that provides a unified view
 * of genes with finished curations across all scopes. This API client only supports
 * read operations.
 */

import apiClient from './client.js'

export const geneCatalogueAPI = {
  /**
   * Get the Gene Catalogue with filtering and pagination.
   *
   * @param {Object} params - Query parameters
   * @param {number} params.skip - Number of records to skip (default: 0)
   * @param {number} params.limit - Maximum records to return (default: 20, max: 100)
   * @param {string} params.search - Search gene symbol, HGNC ID, or aliases
   * @param {string} params.classification - Filter by classification (definitive, strong, etc.)
   * @param {string} params.disease - Search/filter by disease name
   * @param {string} params.scope_id - Filter by specific scope UUID
   * @param {string} params.chromosome - Filter by chromosome
   * @param {string} params.sort_by - Sort field (approved_symbol, total_curations, chromosome)
   * @param {string} params.sort_order - Sort order (asc or desc)
   * @returns {Promise<Object>} Catalogue response with entries, pagination, and summary stats
   */
  async getCatalogue(params = {}) {
    const response = await apiClient.get('/gene-catalogue/', { params })
    return response.data
  },

  /**
   * Get available classifications for filtering.
   *
   * @returns {Promise<string[]>} List of classification values with active curations
   */
  async getAvailableClassifications() {
    const response = await apiClient.get('/gene-catalogue/filters/classifications')
    return response.data
  },

  /**
   * Get available diseases for filtering.
   *
   * @param {number} limit - Maximum diseases to return (default: 50, max: 200)
   * @returns {Promise<string[]>} List of unique disease names
   */
  async getAvailableDiseases(limit = 50) {
    const response = await apiClient.get('/gene-catalogue/filters/diseases', {
      params: { limit }
    })
    return response.data
  },

  /**
   * Get scopes available for filtering.
   *
   * @returns {Promise<Object[]>} List of scopes with active curations
   */
  async getAvailableScopes() {
    const response = await apiClient.get('/gene-catalogue/filters/scopes')
    return response.data
  }
}

File: backend/app/api/v1/endpoints/gene_catalogue.py
"""Gene Catalogue API endpoints.

The Gene Catalogue is a read-only aggregation layer that provides a unified view
of genes with finished curations across all scopes. It exists independently from
scope-level data and never modifies source data.

Key principles:
- Read-only: No write operations
- Scope-independent: Does not affect scope workflows
- Derived data only: All content computed from active curations
- Aggregation by gene: Same gene across scopes consolidated
"""

from uuid import UUID

from fastapi import APIRouter, Depends, Query
from sqlalchemy.orm import Session

from app.core.database import get_db
from app.core.logging import api_endpoint
from app.crud import gene_catalogue as catalogue_crud
from app.schemas.gene_catalogue import (
    GeneCatalogueFilters,
    GeneCatalogueResponse,
)

router = APIRouter()


@router.get("/", response_model=GeneCatalogueResponse)
@api_endpoint()
def get_gene_catalogue(
    db: Session = Depends(get_db),
    skip: int = Query(0, ge=0, description="Number of records to skip"),
    limit: int = Query(20, ge=1, le=100, description="Maximum records to return"),
    search: str | None = Query(
        None, description="Search gene symbol, HGNC ID, or aliases"
    ),
    classification: str | None = Query(
        None, description="Filter by classification (definitive, strong, etc.)"
    ),
    disease: str | None = Query(None, description="Search/filter by disease name"),
    scope_id: UUID | None = Query(None, description="Filter by specific scope"),
    chromosome: str | None = Query(None, description="Filter by chromosome"),
    sort_by: str = Query(
        "approved_symbol",
        description="Sort field: approved_symbol, total_curations, chromosome",
    ),
    sort_order: str = Query("asc", description="Sort order: asc or desc"),
) -> GeneCatalogueResponse:
    """
    Get the Gene Catalogue.

    Returns genes with at least one active (finished) curation across all scopes.
    This is a read-only, public endpoint that aggregates curation data.

    The catalogue shows:
    - Gene identification (symbol, HGNC ID, location)
    - Total curations and scope count per gene
    - Classification distribution across scopes
    - Unique diseases associated with the gene
    - Per-scope curation details (disease, classification, scores)

    Supports filtering by:
    - Gene search (symbol, HGNC ID, aliases)
    - Classification (definitive, strong, moderate, etc.)
    - Disease name (partial match)
    - Specific scope
    - Chromosome

    No authentication required - this is a public reporting endpoint.
    """
    filters = GeneCatalogueFilters(
        search=search,
        classification=classification,
        disease=disease,
        scope_id=scope_id,
        chromosome=chromosome,
        skip=skip,
        limit=limit,
        sort_by=sort_by,
        sort_order=sort_order,
    )

    return catalogue_crud.get_gene_catalogue(db, filters)


@router.get("/filters/classifications", response_model=list[str])
@api_endpoint()
def get_available_classifications(
    db: Session = Depends(get_db),
) -> list[str]:
    """
    Get available classifications for filtering.

    Returns list of classification values that have at least one
    active curation in the catalogue.

    Public endpoint - no authentication required.
    """
    return catalogue_crud.get_available_classifications(db)


@router.get("/filters/diseases", response_model=list[str])
@api_endpoint()
def get_available_diseases(
    db: Session = Depends(get_db),
    limit: int = Query(50, ge=1, le=200, description="Maximum diseases to return"),
) -> list[str]:
    """
    Get available diseases for filtering.

    Returns list of unique disease names from active curations.

    Public endpoint - no authentication required.
    """
    return catalogue_crud.get_available_diseases(db, limit=limit)


@router.get("/filters/scopes", response_model=list[dict[str, str | bool]])
@api_endpoint()
def get_available_scopes(
    db: Session = Depends(get_db),
) -> list[dict[str, str | bool]]:
    """
    Get scopes available for filtering.

    Returns scopes that have at least one active curation.

    Public endpoint - no authentication required.
    """
    return catalogue_crud.get_available_scopes_for_catalogue(db)
